pyquil/operator_estimation.py

```diff
@@ -0,0 +1,428 @@
+    """
+    in_operator: PauliTerm
+    out_operator: PauliTerm
+
+    def __str__(self):
+        return f'{self.in_operator}\u2192{self.out_operator}'
+
+    def __repr__(self):
+        return f'Experiment[{self}]'
+
+    def serializable(self):
+        return str(self)
+
+    @classmethod
+    def from_str(cls, s: str):
+        """The opposite of str(exp)"""
+        instr, outstr = s.split('\u2192')
+        return cls(in_operator=PauliTerm.from_str(instr),
+                   out_operator=PauliTerm.from_str(outstr))
+
+
+def _abbrev_program(program: Program, abbrev_after=10):
+    program_lines = program.out().splitlines()
+    if abbrev_after is not None and len(program_lines) > abbrev_after:
+        first_n = abbrev_after // 2
+        last_n = abbrev_after - first_n
+        excluded = len(program_lines) - abbrev_after
+        program_lines = (program_lines[:first_n] + [f'... {excluded} hidden ...']
+                         + program_lines[-last_n:])
+
+    return '; '.join(program_lines)
+
+
+class ExperimentSuite:
+    """
+    A whole set of tomography-like experiments
+
+    Many near-term quantum algorithms involve measuring an ansatz state with respect to
+    many different observables. This class contains all (in_pauli, out_pauli) pairs for
+    a given ansatz preparation program.
+
+    Experiments belonging to a shared tensor product basis (TPB) can (optionally) be estimated
+    simultaneously. Therefore, this class is backed by a list of list of Experiments.
+    Experiments sharing an inner list will be estimated simultaneously. If you don't want this,
+    provide a list of length-1-lists. As a convenience, if you pass a 1D list to the constructor
+    it will expand it to a list of length-1-lists.
+
+    Please see :py:func:`group_experiments` for a function that will automatically process an
+    ExperimentSuite to group Experiments sharing a TPB.
+    """
+
+    def __init__(self,
+                 experiments: Union[List[Experiment], List[List[Experiment]]],
+                 program: Program,
+                 qubits: List[int]):
+        if len(experiments) == 0:
+            experiments = []
+        else:
+            if isinstance(experiments[0], Experiment):
+                # convenience wrapping in lists of length 1
+                experiments = [[exp] for exp in experiments]
+
+        self._experiments = experiments  # type: List[List[Experiment]]
+        self.program = program
+        self.qubits = qubits
+
+    def __len__(self):
+        return len(self._experiments)
+
+    def __getitem__(self, item):
+        return self._experiments[item]
+
+    def __setitem__(self, key, value):
+        self._experiments[key] = value
+
+    def __delitem__(self, key):
+        self._experiments.__delitem__(key)
+
+    def __iter__(self):
+        yield from self._experiments
+
+    def __reversed__(self):
+        yield from reversed(self._experiments)
+
+    def __contains__(self, item):
```

Comment:
Somewhat relates to my comment above about the `__iter__` and `__reversed__` functions. With this definition of `__contains__`, an Experiment is never contained in an ExperimentSuite. If we do
>>> import pyquil.operator_estimation as oe
>>> import pyquil.paulis as pl
>>> expt1 = oe.Experiment(pl.sX(1) * pl.sZ(0), pl.sX(2) * pl.sY(1) * pl.sZ(0))
>>> expt2 = oe.Experiment(pl.sZ(1) * pl.sX(0), pl.sY(2) * pl.sZ(1) * pl.sX(0))
>>> expt3 = oe.Experiment(pl.sX(2) * pl.sY(1) * pl.sZ(0), pl.sZ(2) * pl.sX(1) * pl.sY(0))
>>> expt_list = [expt1, expt2, expt3]
>>> expt_suite = oe.ExperimentSuite(expt_list, Program(H(0), H(1), H(2)), [0, 1, 2])
And then do
>>> expt1 in expt_suite
We’ll get “False” as output, which isn’t very nice. Note: we’ll get “True” if we do [expt1] in expt_suite.

Reply:
ah yes this is really annoying. would it be terrible to have `__iter__` and `__reversed__` still iterate over the groups (lists) of experiments but modify `__contains__` to operate on a flattened list of experiments?

Reply:
Reiterating a comment I made elsewhere: given that term grouping is an important thing we'll do with instances of `ExperimentSuite`, and given that different term grouping methods can give different groups, we might be interested in checking whether a given group of `Experiment`s exists in an `ExperimentSuite`.

At the same time, we might also be interested in checking whether some single `Experiment` exists in that particular `ExperimentSuite` at all. In such a case, an obvious first guess might be to try something like `[expt] in expt_suite` (instead of `expt in expt_suite`) but which might return `False` if that `expt` is grouped with some other `Experiment`s.

So what could be better is if we added something like the following to the doc-strings:
```
If checking for whether an `Experiment` exists in an `ExperimentSuite`, please use the `Experiment in ExperimentSuite` format. If checking for whether a group of `Experiment`s exists in an `ExperimentSuite`, please use the `[ExperimentA, ExperimentB, ExperimentC, ...] in ExperimentSuite` (where `Experiment`s `A-C` are all instances of the `Experiment` class)
```
and then modified the code to something like the following (the function for flattening borrowed from the awesome blog post at http://rightfootin.blogspot.com/2006/09/more-on-python-flatten.html; this might be overkill, since we really only need to flatten nested-level-2 lists in this case)
```
def flattened(l, ltypes=(list, tuple)):
    ltype = type(l)
    l = list(l)
    i = 0
    while i < len(l):
        while isinstance(l[i], ltypes):
            if not l[i]:
                l.pop(i)
                i -= 1
                break
            else:
                l[i:i + 1] = l[i]
        i += 1
    return ltype(l)


def ___contains___(self, item):
    return item in self._experiments or item in flattened(self._experiments)
```